File: leetcode/3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.cpp
// Last updated: 29/04/2026, 23:14:40
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int maxLength = 0;
        int i = 0, j = 0;
        unordered_set<char> visited;

        while (j < s.size()) {
            if (visited.find(s[j]) != visited.end()) {
                visited.erase(s[i]);
                i++;
            } else {
                visited.insert(s[j]);
                maxLength = max(maxLength, int(visited.size()));
                j++;
            }
        }
        return maxLength;
    }
};

File: leetcode/3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.java
// Last updated: 29/04/2026, 23:14:37
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int maxR = 0;
        Map<Character, Integer> map = new HashMap<>();
        for(int i=0; i<s.length();) {
            Set<Character> set = new HashSet<>();
            int end = i;
            int currLen = 0;
            while(end<s.length() &&!set.contains(s.charAt(end))) {
                set.add(s.charAt(end));
                map.put(s.charAt(end), end);
                currLen+=1;
                end+=1;
            }
            if(end<s.length()) {
                i = map.get(s.charAt(end))+1;
            } else {
                i++;
            }
            maxR = Math.max(currLen, maxR);
        }
        return maxR;
    }
}